Log network calls

abelgardep opened the pull request:
Implements https://github.com/owncloud/android/issues/2887
Needs https://github.com/owncloud/android-library/pull/345

____

QA

Test plan: https://github.com/owncloud/QA/blob/master/Mobile/Android/Release_2.16/2981-logging_network.md

BUGS & IMPROVEMENTS:

- [X] (1) Body not logged https://github.com/owncloud/android/pull/2981#issuecomment-712629210 [FIXED]
- [X] (2) Request and responses are the same https://github.com/owncloud/android/pull/2981#issuecomment-712632175 [FIXED]
- [X] (3) Log entries mixed https://github.com/owncloud/android/pull/2981#issuecomment-712634328 [FIXED]
- [x] (4) Release version does not make the section appear in live https://github.com/owncloud/android/pull/2981#issuecomment-712654891 [FIXED]
- [X] (5) Response body marked as request body https://github.com/owncloud/android/pull/2981#issuecomment-713662277
- [x] (6) Error in uploads https://github.com/owncloud/android/pull/2981#issuecomment-714380075 [FIXED]
- [X] (7) Removing logs from binary bodys. Typo. https://github.com/owncloud/android/pull/2981#issuecomment-717243183

abelgardep commented:
Moving this forward, i need some feedback about it. @michaelstingl @theScrabi @jesmrec

michaelstingl commented:
> Moving this forward, i need some feedback about it. @michaelstingl @theScrabi @jesmrec

Sorry, what is the exact question? Seen https://github.com/owncloud/android/issues/2887#issuecomment-703518689 ?

abelgardep commented:
I mean about the current implementation. I have added a new log option. And http logs are like this:

![Screenshot 2020-10-07 at 11 21 42](https://user-images.githubusercontent.com/47524927/95312769-9f3f3f80-088f-11eb-9b54-a59f355f82fc.png)


`D/(LogBuilder.kt:33) .logHttp(): [Network, request] [info] Type: GET URL: https://demo.owncloud.com/index.php/avatar/demo/384
D/(LogBuilder.kt:33) .logHttp(): [Network, request] [header] (Cache-Control, no-cache, no-store, must-revalidate)
D/(LogBuilder.kt:33) .logHttp(): [Network, request] [header] (Content-Length, 31)
D/(LogBuilder.kt:33) .logHttp(): [Network, request] [header] (Content-Security-Policy, default-src 'none';manifest-src 'self';script-src 'self' 'unsafe-eval';style-src 'self' 'unsafe-inline';img-src 'self' data: blob:;font-src 'self';connect-src 'self';media-src 'self')
D/(LogBuilder.kt:33) .logHttp(): [Network, request] [header] (Content-Type, application/json; charset=utf-8)
D/(LogBuilder.kt:33) .logHttp(): [Network, request] [header] (Date, Wed, 07 Oct 2020 09:21:46 GMT)
D/(LogBuilder.kt:33) .logHttp(): [Network, request] [header] (Expires, Thu, 19 Nov 1981 08:52:00 GMT)
D/(LogBuilder.kt:33) .logHttp(): [Network, request] [header] (Last-Modified, Wed, 07 Oct 2020 09:21:46 +0000)
D/(LogBuilder.kt:33) .logHttp(): [Network, request] [header] (Pragma, public)
D/(LogBuilder.kt:33) .logHttp(): [Network, request] [header] (Referrer-Policy, strict-origin-when-cross-origin)
D/(LogBuilder.kt:33) .logHttp(): [Network, request] [header] (Server, Apache)
D/(LogBuilder.kt:33) .logHttp(): [Network, request] [header] (Set-Cookie, occhdt1yw3ys=sbe9g0rlip138fdljp40nv5mug; path=/; secure; HttpOnly)
D/(LogBuilder.kt:33) .logHttp(): [Network, request] [header] (Strict-Transport-Security, max-age=315360000; preload)
D/(LogBuilder.kt:33) .logHttp(): [Network, request] [header] (X-Content-Type-Options, nosniff)
D/(LogBuilder.kt:33) .logHttp(): [Network, request] [header] (X-Download-Options, noopen)
D/(LogBuilder.kt:33) .logHttp(): [Network, request] [header] (X-Frame-Options, SAMEORIGIN)
D/(LogBuilder.kt:33) .logHttp(): [Network, request] [header] (X-Permitted-Cross-Domain-Policies, none)
D/(LogBuilder.kt:33) .logHttp(): [Network, request] [header] (X-Robots-Tag, none)
D/(LogBuilder.kt:33) .logHttp(): [Network, request] [header] (X-Xss-Protection, 1; mode=block)
D/(LogBuilder.kt:33) .logHttp(): [Network, request] [body] okhttp3.internal.http.RealResponseBody@8389c96
D/(LogBuilder.kt:33) .logHttp(): [Network, response] [info] Code: 200  Message: OK IsSuccessful: true
D/(LogBuilder.kt:33) .logHttp(): [Network, response] [header] (Cache-Control, no-cache, no-store, must-revalidate)
D/(LogBuilder.kt:33) .logHttp(): [Network, response] [header] (Content-Length, 31)
D/(LogBuilder.kt:33) .logHttp(): [Network, response] [header] (Content-Security-Policy, default-src 'none';manifest-src 'self';script-src 'self' 'unsafe-eval';style-src 'self' 'unsafe-inline';img-src 'self' data: blob:;font-src 'self';connect-src 'self';media-src 'self')
D/(LogBuilder.kt:33) .logHttp(): [Network, response] [header] (Content-Type, application/json; charset=utf-8)
D/(LogBuilder.kt:33) .logHttp(): [Network, response] [header] (Date, Wed, 07 Oct 2020 09:21:46 GMT)
D/(LogBuilder.kt:33) .logHttp(): [Network, response] [header] (Expires, Thu, 19 Nov 1981 08:52:00 GMT)
D/(LogBuilder.kt:33) .logHttp(): [Network, response] [header] (Last-Modified, Wed, 07 Oct 2020 09:21:46 +0000)
D/(LogBuilder.kt:33) .logHttp(): [Network, response] [header] (Pragma, public)
D/(LogBuilder.kt:33) .logHttp(): [Network, response] [header] (Referrer-Policy, strict-origin-when-cross-origin)
D/(LogBuilder.kt:33) .logHttp(): [Network, response] [header] (Server, Apache)
D/(LogBuilder.kt:33) .logHttp(): [Network, response] [header] (Set-Cookie, occhdt1yw3ys=sbe9g0rlip138fdljp40nv5mug; path=/; secure; HttpOnly)
D/(LogBuilder.kt:33) .logHttp(): [Network, response] [header] (Strict-Transport-Security, max-age=315360000; preload)
D/(LogBuilder.kt:33) .logHttp(): [Network, response] [header] (X-Content-Type-Options, nosniff)
D/(LogBuilder.kt:33) .logHttp(): [Network, response] [header] (X-Download-Options, noopen)
D/(LogBuilder.kt:33) .logHttp(): [Network, response] [header] (X-Frame-Options, SAMEORIGIN)
D/(LogBuilder.kt:33) .logHttp(): [Network, response] [header] (X-Permitted-Cross-Domain-Policies, none)
D/(LogBuilder.kt:33) .logHttp(): [Network, response] [header] (X-Robots-Tag, none)
D/(LogBuilder.kt:33) .logHttp(): [Network, response] [header] (X-Xss-Protection, 1; mode=block)
D/(LogBuilder.kt:33) .logHttp(): [Network, response] [body] okhttp3.internal.http.RealResponseBody@8389c96`

theScrabi reviewed (changes requested):
LFTM, but please check your commit names, there are spelling erros, and some names have a different meaning or commits could be squashed.

michaelstingl reviewed (changes requested):
Add a few lines to the troubleshooting docs?
https://github.com/owncloud/android/blob/master/docs/modules/ROOT/pages/troubleshooting.adoc

abelgardep commented:
> Add a few lines to the troubleshooting docs?
> https://github.com/owncloud/android/blob/master/docs/modules/ROOT/pages/troubleshooting.adoc

Done

jesmrec commented:
### (1)

Body requests and responses are not logged. This is the text logged instead:

```D: 2020-10-19 15:01:02:884(LogBuilder.kt:33) .logHttp()[Network, request] [body]
okhttp3.internal.http.RealResponseBody@790056e
```

jesmrec commented:
### (2)

Logging requests and reponses are duplicated, logging only the responses. For example:

```
D: 2020-10-19 15:01:03:628(LogBuilder.kt:33) .logHttp()[Network, request] [info] Type: PROPFIND URL: http://192.168.1.66:33000/remote.php/dav/files/user1
D: 2020-10-19 15:01:03:638(LogBuilder.kt:33) .logHttp()[Network, request] [header] (Date, Mon, 19 Oct 2020 13:01:03 GMT)
D: 2020-10-19 15:01:03:644(LogBuilder.kt:33) .logHttp()[Network, request] [header] (Server, Apache)
D: 2020-10-19 15:01:03:649(LogBuilder.kt:33) .logHttp()[Network, request] [header] (X-Content-Type-Options, nosniff)
D: 2020-10-19 15:01:03:655(LogBuilder.kt:33) .logHttp()[Network, request] [header] (X-XSS-Protection, 1; mode=block)
D: 2020-10-19 15:01:03:662(LogBuilder.kt:33) .logHttp()[Network, request] [header] (X-Robots-Tag, none)
D: 2020-10-19 15:01:03:667(LogBuilder.kt:33) .logHttp()[Network, request] [header] (X-Frame-Options, SAMEORIGIN)
D: 2020-10-19 15:01:03:671(LogBuilder.kt:33) .logHttp()[Network, request] [header] (X-Download-Options, noopen)
D: 2020-10-19 15:01:03:674(LogBuilder.kt:33) .logHttp()[Network, request] [header] (X-Permitted-Cross-Domain-Policies, none)
D: 2020-10-19 15:01:03:677(LogBuilder.kt:33) .logHttp()[Network, request] [header] (Expires, Thu, 19 Nov 1981 08:52:00 GMT)
D: 2020-10-19 15:01:03:686(LogBuilder.kt:33) .logHttp()[Network, request] [header] (Cache-Control, no-store, no-cache, must-revalidate)
D: 2020-10-19 15:01:03:689(LogBuilder.kt:33) .logHttp()[Network, request] [header] (Pragma, no-cache)
D: 2020-10-19 15:01:03:692(LogBuilder.kt:33) .logHttp()[Network, request] [header] (Content-Security-Policy, default-src 'none';)
D: 2020-10-19 15:01:03:695(LogBuilder.kt:33) .logHttp()[Network, request] [header] (Vary, Brief,Prefer)
D: 2020-10-19 15:01:03:698(LogBuilder.kt:33) .logHttp()[Network, request] [header] (DAV, 1, 3, extended-mkcol, 2)
D: 2020-10-19 15:01:03:701(LogBuilder.kt:33) .logHttp()[Network, request] [header] (Content-Length, 387)
D: 2020-10-19 15:01:03:705(LogBuilder.kt:33) .logHttp()[Network, request] [header] (Keep-Alive, timeout=5, max=98)
D: 2020-10-19 15:01:03:708(LogBuilder.kt:33) .logHttp()[Network, request] [header] (Connection, Keep-Alive)
D: 2020-10-19 15:01:03:710(LogBuilder.kt:33) .logHttp()[Network, request] [header] (Content-Type, application/xml; charset=utf-8)
D: 2020-10-19 15:01:03:714(LogBuilder.kt:33) .logHttp()[Network, request] [body] okhttp3.internal.http.RealResponseBody@3c130cb
D: 2020-10-19 15:01:03:716(LogBuilder.kt:33) .logHttp()[Network, response] [info] Code: 207  Message: Multi-Status IsSuccessful: true
D: 2020-10-19 15:01:03:717(LogBuilder.kt:33) .logHttp()[Network, response] [header] (Date, Mon, 19 Oct 2020 13:01:03 GMT)
D: 2020-10-19 15:01:03:719(LogBuilder.kt:33) .logHttp()[Network, response] [header] (Server, Apache)
D: 2020-10-19 15:01:03:721(LogBuilder.kt:33) .logHttp()[Network, response] [header] (X-Content-Type-Options, nosniff)
D: 2020-10-19 15:01:03:722(LogBuilder.kt:33) .logHttp()[Network, response] [header] (X-XSS-Protection, 1; mode=block)
D: 2020-10-19 15:01:03:724(LogBuilder.kt:33) .logHttp()[Network, response] [header] (X-Robots-Tag, none)
D: 2020-10-19 15:01:03:726(LogBuilder.kt:33) .logHttp()[Network, response] [header] (X-Frame-Options, SAMEORIGIN)
D: 2020-10-19 15:01:03:728(LogBuilder.kt:33) .logHttp()[Network, response] [header] (X-Download-Options, noopen)
D: 2020-10-19 15:01:03:730(LogBuilder.kt:33) .logHttp()[Network, response] [header] (X-Permitted-Cross-Domain-Policies, none)
D: 2020-10-19 15:01:03:731(LogBuilder.kt:33) .logHttp()[Network, response] [header] (Expires, Thu, 19 Nov 1981 08:52:00 GMT)
D: 2020-10-19 15:01:03:733(LogBuilder.kt:33) .logHttp()[Network, response] [header] (Cache-Control, no-store, no-cache, must-revalidate)
D: 2020-10-19 15:01:03:734(LogBuilder.kt:33) .logHttp()[Network, response] [header] (Pragma, no-cache)
D: 2020-10-19 15:01:03:736(LogBuilder.kt:33) .logHttp()[Network, response] [header] (Content-Security-Policy, default-src 'none';)
D: 2020-10-19 15:01:03:738(LogBuilder.kt:33) .logHttp()[Network, response] [header] (Vary, Brief,Prefer)
D: 2020-10-19 15:01:03:739(LogBuilder.kt:33) .logHttp()[Network, response] [header] (DAV, 1, 3, extended-mkcol, 2)
D: 2020-10-19 15:01:03:741(LogBuilder.kt:33) .logHttp()[Network, response] [header] (Content-Length, 387)
D: 2020-10-19 15:01:03:743(LogBuilder.kt:33) .logHttp()[Network, response] [header] (Keep-Alive, timeout=5, max=98)
D: 2020-10-19 15:01:03:748(LogBuilder.kt:33) .logHttp()[Network, response] [header] (Connection, Keep-Alive)
D: 2020-10-19 15:01:03:750(LogBuilder.kt:33) .logHttp()[Network, response] [header] (Content-Type, application/xml; charset=utf-8)
D: 2020-10-19 15:01:03:752(LogBuilder.kt:33) .logHttp()[Network, response] [body] okhttp3.internal.http.RealResponseBody@3c130cb
```

the set of headers marked as `[Network, request]` is the same as `[Network, response]`

jesmrec commented:
### (3)

Not sure if this is a problem or a feature: i have seen that sometimes two requests overlaps in the logs:

```
D: 2020-10-19 15:01:03:262(LogBuilder.kt:33) .logHttp()[Network, request] [info] Type: PROPFIND URL: http://192.168.1.66:33000/remote.php/dav/files/user1/
D: 2020-10-19 15:01:03:263(LogBuilder.kt:33) .logHttp()[Network, request] [info] Type: GET URL: http://192.168.1.66:33000/ocs/v2.php/cloud/user?format=json
D: 2020-10-19 15:01:03:265(LogBuilder.kt:33) .logHttp()[Network, request] [header] (Date, Mon, 19 Oct 2020 13:01:03 GMT)
D: 2020-10-19 15:01:03:265(LogBuilder.kt:33) .logHttp()[Network, request] [header] (Date, Mon, 19 Oct 2020 13:01:03 GMT)
D: 2020-10-19 15:01:03:266(LogBuilder.kt:33) .logHttp()[Network, request] [header] (Server, Apache)
D: 2020-10-19 15:01:03:266(LogBuilder.kt:33) .logHttp()[Network, request] [header] (Server, Apache)
D: 2020-10-19 15:01:03:268(LogBuilder.kt:33) .logHttp()[Network, request] [header] (X-Content-Type-Options, nosniff)
D: 2020-10-19 15:01:03:268(LogBuilder.kt:33) .logHttp()[Network, request] [header] (X-Content-Type-Options, nosniff)
D: 2020-10-19 15:01:03:270(LogBuilder.kt:33) .logHttp()[Network, request] [header] (X-XSS-Protection, 1; mode=block)
D: 2020-10-19 15:01:03:270(LogBuilder.kt:33) .logHttp()[Network, request] [header] (X-XSS-Protection, 1; mode=block)
D: 2020-10-19 15:01:03:272(LogBuilder.kt:33) .logHttp()[Network, request] [header] (X-Robots-Tag, none)
D: 2020-10-19 15:01:03:272(LogBuilder.kt:33) .logHttp()[Network, request] [header] (X-Robots-Tag, none)
D: 2020-10-19 15:01:03:274(LogBuilder.kt:33) .logHttp()[Network, request] [header] (X-Frame-Options, SAMEORIGIN)
D: 2020-10-19 15:01:03:274(LogBuilder.kt:33) .logHttp()[Network, request] [header] (X-Frame-Options, SAMEORIGIN)
D: 2020-10-19 15:01:03:275(LogBuilder.kt:33) .logHttp()[Network, request] [header] (X-Download-Options, noopen)
D: 2020-10-19 15:01:03:275(LogBuilder.kt:33) .logHttp()[Network, request] [header] (X-Download-Options, noopen)
D: 2020-10-19 15:01:03:277(LogBuilder.kt:33) .logHttp()[Network, request] [header] (X-Permitted-Cross-Domain-Policies, none)
D: 2020-10-19 15:01:03:277(LogBuilder.kt:33) .logHttp()[Network, request] [header] (X-Permitted-Cross-Domain-Policies, none)
D: 2020-10-19 15:01:03:278(LogBuilder.kt:33) .logHttp()[Network, request] [header] (Expires, Thu, 19 Nov 1981 08:52:00 GMT)
D: 2020-10-19 15:01:03:279(LogBuilder.kt:33) .logHttp()[Network, request] [header] (Expires, Thu, 19 Nov 1981 08:52:00 GMT)
D: 2020-10-19 15:01:03:280(LogBuilder.kt:33) .logHttp()[Network, request] [header] (Cache-Control, no-store, no-cache, must-revalidate)
D: 2020-10-19 15:01:03:280(LogBuilder.kt:33) .logHttp()[Network, request] [header] (Cache-Control, no-cache, no-store, must-revalidate)
D: 2020-10-19 15:01:03:282(LogBuilder.kt:33) .logHttp()[Network, request] [header] (Pragma, no-cache)
D: 2020-10-19 15:01:03:282(LogBuilder.kt:33) .logHttp()[Network, request] [header] (Pragma, no-cache)
D: 2020-10-19 15:01:03:283(LogBuilder.kt:33) .logHttp()[Network, request] [header] (Content-Security-Policy, default-src 'none';)
D: 2020-10-19 15:01:03:283(LogBuilder.kt:33) .logHttp()[Network, request] [header] (Content-Security-Policy, default-src 'none';manifest-src 'self';script-src 'self' 'unsafe-eval';style-src 'self' 'unsafe-inline';img-src 'self' data: blob:;font-src 'self';connect-src 'self';media-src 'self')
D: 2020-10-19 15:01:03:285(LogBuilder.kt:33) .logHttp()[Network, request] [header] (Vary, Brief,Prefer)
D: 2020-10-19 15:01:03:285(LogBuilder.kt:33) .logHttp()[Network, request] [header] (Content-Length, 156)
D: 2020-10-19 15:01:03:286(LogBuilder.kt:33) .logHttp()[Network, request] [header] (DAV, 1, 3, extended-mkcol, 2)
D: 2020-10-19 15:01:03:286(LogBuilder.kt:33) .logHttp()[Network, request] [header] (Keep-Alive, timeout=5, max=100)
D: 2020-10-19 15:01:03:287(LogBuilder.kt:33) .logHttp()[Network, request] [header] (Keep-Alive, timeout=5, max=99)
D: 2020-10-19 15:01:03:288(LogBuilder.kt:33) .logHttp()[Network, request] [header] (Connection, Keep-Alive)
D: 2020-10-19 15:01:03:289(LogBuilder.kt:33) .logHttp()[Network, request] [header] (Connection, Keep-Alive)
D: 2020-10-19 15:01:03:289(LogBuilder.kt:33) .logHttp()[Network, request] [header] (Content-Type, application/json; charset=utf-8)
D: 2020-10-19 15:01:03:291(LogBuilder.kt:33) .logHttp()[Network, request] [header] (Transfer-Encoding, chunked)
D: 2020-10-19 15:01:03:291(LogBuilder.kt:33) .logHttp()[Network, request] [body] okhttp3.internal.http.RealResponseBody@d3576a2
D: 2020-10-19 15:01:03:292(LogBuilder.kt:33) .logHttp()[Network, request] [header] (Content-Type, application/xml; charset=utf-8)
D: 2020-10-19 15:01:03:292(LogBuilder.kt:33) .logHttp()[Network, response] [info] Code: 200  Message: OK IsSuccessful: true
D: 2020-10-19 15:01:03:294(LogBuilder.kt:33) .logHttp()[Network, request] [body] okhttp3.internal.http.RealResponseBody@bb93a25
D: 2020-10-19 15:01:03:294(LogBuilder.kt:33) .logHttp()[Network, response] [header] (Date, Mon, 19 Oct 2020 13:01:03 GMT)
D: 2020-10-19 15:01:03:295(LogBuilder.kt:33) .logHttp()[Network, response] [info] Code: 207  Message: Multi-Status IsSuccessful: true
D: 2020-10-19 15:01:03:295(LogBuilder.kt:33) .logHttp()[Network, response] [header] (Server, Apache)
D: 2020-10-19 15:01:03:296(LogBuilder.kt:33) .logHttp()[Network, response] [header] (X-Content-Type-Options, nosniff)
D: 2020-10-19 15:01:03:296(LogBuilder.kt:33) .logHttp()[Network, response] [header] (Date, Mon, 19 Oct 2020 13:01:03 GMT)
D: 2020-10-19 15:01:03:297(LogBuilder.kt:33) .logHttp()[Network, response] [header] (Server, Apache)
D: 2020-10-19 15:01:03:297(LogBuilder.kt:33) .logHttp()[Network, response] [header] (X-XSS-Protection, 1; mode=block)
D: 2020-10-19 15:01:03:299(LogBuilder.kt:33) .logHttp()[Network, response] [header] (X-Content-Type-Options, nosniff)
D: 2020-10-19 15:01:03:299(LogBuilder.kt:33) .logHttp()[Network, response] [header] (X-Robots-Tag, none)
D: 2020-10-19 15:01:03:300(LogBuilder.kt:33) .logHttp()[Network, response] [header] (X-XSS-Protection, 1; mode=block)
D: 2020-10-19 15:01:03:300(LogBuilder.kt:33) .logHttp()[Network, response] [header] (X-Frame-Options, SAMEORIGIN)
D: 2020-10-19 15:01:03:301(LogBuilder.kt:33) .logHttp()[Network, response] [header] (X-Robots-Tag, none)
D: 2020-10-19 15:01:03:301(LogBuilder.kt:33) .logHttp()[Network, response] [header] (X-Download-Options, noopen)
D: 2020-10-19 15:01:03:302(LogBuilder.kt:33) .logHttp()[Network, response] [header] (X-Permitted-Cross-Domain-Policies, none)
D: 2020-10-19 15:01:03:302(LogBuilder.kt:33) .logHttp()[Network, response] [header] (X-Frame-Options, SAMEORIGIN)
D: 2020-10-19 15:01:03:303(LogBuilder.kt:33) .logHttp()[Network, response] [header] (X-Download-Options, noopen)
D: 2020-10-19 15:01:03:303(LogBuilder.kt:33) .logHttp()[Network, response] [header] (Expires, Thu, 19 Nov 1981 08:52:00 GMT)
D: 2020-10-19 15:01:03:304(LogBuilder.kt:33) .logHttp()[Network, response] [header] (Cache-Control, no-cache, no-store, must-revalidate)
D: 2020-10-19 15:01:03:304(LogBuilder.kt:33) .logHttp()[Network, response] [header] (X-Permitted-Cross-Domain-Policies, none)
D: 2020-10-19 15:01:03:306(LogBuilder.kt:33) .logHttp()[Network, response] [header] (Expires, Thu, 19 Nov 1981 08:52:00 GMT)
D: 2020-10-19 15:01:03:308(LogBuilder.kt:33) .logHttp()[Network, response] [header] (Pragma, no-cache)
D: 2020-10-19 15:01:03:308(LogBuilder.kt:33) .logHttp()[Network, response] [header] (Cache-Control, no-store, no-cache, must-revalidate)
D: 2020-10-19 15:01:03:309(LogBuilder.kt:33) .logHttp()[Network, response] [header] (Pragma, no-cache)
D: 2020-10-19 15:01:03:309(LogBuilder.kt:33) .logHttp()[Network, response] [header] (Content-Security-Policy, default-src 'none';manifest-src 'self';script-src 'self' 'unsafe-eval';style-src 'self' 'unsafe-inline';img-src 'self' data: blob:;font-src 'self';connect-src 'self';media-src 'self')
D: 2020-10-19 15:01:03:311(LogBuilder.kt:33) .logHttp()[Network, response] [header] (Content-Security-Policy, default-src 'none';)
D: 2020-10-19 15:01:03:311(LogBuilder.kt:33) .logHttp()[Network, response] [header] (Content-Length, 156)
D: 2020-10-19 15:01:03:312(LogBuilder.kt:33) .logHttp()[Network, response] [header] (Keep-Alive, timeout=5, max=100)
D: 2020-10-19 15:01:03:312(LogBuilder.kt:33) .logHttp()[Network, response] [header] (Vary, Brief,Prefer)
D: 2020-10-19 15:01:03:313(LogBuilder.kt:33) .logHttp()[Network, response] [header] (Connection, Keep-Alive)
D: 2020-10-19 15:01:03:313(LogBuilder.kt:33) .logHttp()[Network, response] [header] (DAV, 1, 3, extended-mkcol, 2)
D: 2020-10-19 15:01:03:314(LogBuilder.kt:33) .logHttp()[Network, response] [header] (Keep-Alive, timeout=5, max=99)
D: 2020-10-19 15:01:03:314(LogBuilder.kt:33) .logHttp()[Network, response] [header] (Content-Type, application/json; charset=utf-8)
D: 2020-10-19 15:01:03:315(LogBuilder.kt:33) .logHttp()[Network, response] [header] (Connection, Keep-Alive)
D: 2020-10-19 15:01:03:316(LogBuilder.kt:33) .logHttp()[Network, response] [body] okhttp3.internal.http.RealResponseBody@d3576a2
D: 2020-10-19 15:01:03:317(LogBuilder.kt:33) .logHttp()[Network, response] [header] (Transfer-Encoding, chunked)

```

Here, all request headers and response headers are mixed. Probably this is a threading issue.... just to know if something con be done.

jesmrec commented:
### (4)

1. Build a release apk
2. Add an account and enter in `Settings` -> Log section is hidden
3. Tap 5 times on build number to unlock

Current: Log section does not appear. You have to go out and enter again in `Settings` to make it visible
Expected: Live appearance (is it feasible? in the `master` version, `Logs`are displayed just after the 5th click )

abelgardep commented:
> ### (3)
> Not sure if this is a problem or a feature: i have seen that sometimes two requests overlaps in the logs:
> 
> ```
> D: 2020-10-19 15:01:03:262(LogBuilder.kt:33) .logHttp()[Network, request] [info] Type: PROPFIND URL: http://192.168.1.66:33000/remote.php/dav/files/user1/
> D: 2020-10-19 15:01:03:263(LogBuilder.kt:33) .logHttp()[Network, request] [info] Type: GET URL: http://192.168.1.66:33000/ocs/v2.php/cloud/user?format=json
> D: 2020-10-19 15:01:03:265(LogBuilder.kt:33) .logHttp()[Network, request] [header] (Date, Mon, 19 Oct 2020 13:01:03 GMT)
> D: 2020-10-19 15:01:03:265(LogBuilder.kt:33) .logHttp()[Network, request] [header] (Date, Mon, 19 Oct 2020 13:01:03 GMT)
> D: 2020-10-19 15:01:03:266(LogBuilder.kt:33) .logHttp()[Network, request] [header] (Server, Apache)
> D: 2020-10-19 15:01:03:266(LogBuilder.kt:33) .logHttp()[Network, request] [header] (Server, Apache)
> D: 2020-10-19 15:01:03:268(LogBuilder.kt:33) .logHttp()[Network, request] [header] (X-Content-Type-Options, nosniff)
> D: 2020-10-19 15:01:03:268(LogBuilder.kt:33) .logHttp()[Network, request] [header] (X-Content-Type-Options, nosniff)
> D: 2020-10-19 15:01:03:270(LogBuilder.kt:33) .logHttp()[Network, request] [header] (X-XSS-Protection, 1; mode=block)
> D: 2020-10-19 15:01:03:270(LogBuilder.kt:33) .logHttp()[Network, request] [header] (X-XSS-Protection, 1; mode=block)
> D: 2020-10-19 15:01:03:272(LogBuilder.kt:33) .logHttp()[Network, request] [header] (X-Robots-Tag, none)
> D: 2020-10-19 15:01:03:272(LogBuilder.kt:33) .logHttp()[Network, request] [header] (X-Robots-Tag, none)
> D: 2020-10-19 15:01:03:274(LogBuilder.kt:33) .logHttp()[Network, request] [header] (X-Frame-Options, SAMEORIGIN)
> D: 2020-10-19 15:01:03:274(LogBuilder.kt:33) .logHttp()[Network, request] [header] (X-Frame-Options, SAMEORIGIN)
> D: 2020-10-19 15:01:03:275(LogBuilder.kt:33) .logHttp()[Network, request] [header] (X-Download-Options, noopen)
> D: 2020-10-19 15:01:03:275(LogBuilder.kt:33) .logHttp()[Network, request] [header] (X-Download-Options, noopen)
> D: 2020-10-19 15:01:03:277(LogBuilder.kt:33) .logHttp()[Network, request] [header] (X-Permitted-Cross-Domain-Policies, none)
> D: 2020-10-19 15:01:03:277(LogBuilder.kt:33) .logHttp()[Network, request] [header] (X-Permitted-Cross-Domain-Policies, none)
> D: 2020-10-19 15:01:03:278(LogBuilder.kt:33) .logHttp()[Network, request] [header] (Expires, Thu, 19 Nov 1981 08:52:00 GMT)
> D: 2020-10-19 15:01:03:279(LogBuilder.kt:33) .logHttp()[Network, request] [header] (Expires, Thu, 19 Nov 1981 08:52:00 GMT)
> D: 2020-10-19 15:01:03:280(LogBuilder.kt:33) .logHttp()[Network, request] [header] (Cache-Control, no-store, no-cache, must-revalidate)
> D: 2020-10-19 15:01:03:280(LogBuilder.kt:33) .logHttp()[Network, request] [header] (Cache-Control, no-cache, no-store, must-revalidate)
> D: 2020-10-19 15:01:03:282(LogBuilder.kt:33) .logHttp()[Network, request] [header] (Pragma, no-cache)
> D: 2020-10-19 15:01:03:282(LogBuilder.kt:33) .logHttp()[Network, request] [header] (Pragma, no-cache)
> D: 2020-10-19 15:01:03:283(LogBuilder.kt:33) .logHttp()[Network, request] [header] (Content-Security-Policy, default-src 'none';)
> D: 2020-10-19 15:01:03:283(LogBuilder.kt:33) .logHttp()[Network, request] [header] (Content-Security-Policy, default-src 'none';manifest-src 'self';script-src 'self' 'unsafe-eval';style-src 'self' 'unsafe-inline';img-src 'self' data: blob:;font-src 'self';connect-src 'self';media-src 'self')
> D: 2020-10-19 15:01:03:285(LogBuilder.kt:33) .logHttp()[Network, request] [header] (Vary, Brief,Prefer)
> D: 2020-10-19 15:01:03:285(LogBuilder.kt:33) .logHttp()[Network, request] [header] (Content-Length, 156)
> D: 2020-10-19 15:01:03:286(LogBuilder.kt:33) .logHttp()[Network, request] [header] (DAV, 1, 3, extended-mkcol, 2)
> D: 2020-10-19 15:01:03:286(LogBuilder.kt:33) .logHttp()[Network, request] [header] (Keep-Alive, timeout=5, max=100)
> D: 2020-10-19 15:01:03:287(LogBuilder.kt:33) .logHttp()[Network, request] [header] (Keep-Alive, timeout=5, max=99)
> D: 2020-10-19 15:01:03:288(LogBuilder.kt:33) .logHttp()[Network, request] [header] (Connection, Keep-Alive)
> D: 2020-10-19 15:01:03:289(LogBuilder.kt:33) .logHttp()[Network, request] [header] (Connection, Keep-Alive)
> D: 2020-10-19 15:01:03:289(LogBuilder.kt:33) .logHttp()[Network, request] [header] (Content-Type, application/json; charset=utf-8)
> D: 2020-10-19 15:01:03:291(LogBuilder.kt:33) .logHttp()[Network, request] [header] (Transfer-Encoding, chunked)
> D: 2020-10-19 15:01:03:291(LogBuilder.kt:33) .logHttp()[Network, request] [body] okhttp3.internal.http.RealResponseBody@d3576a2
> D: 2020-10-19 15:01:03:292(LogBuilder.kt:33) .logHttp()[Network, request] [header] (Content-Type, application/xml; charset=utf-8)
> D: 2020-10-19 15:01:03:292(LogBuilder.kt:33) .logHttp()[Network, response] [info] Code: 200  Message: OK IsSuccessful: true
> D: 2020-10-19 15:01:03:294(LogBuilder.kt:33) .logHttp()[Network, request] [body] okhttp3.internal.http.RealResponseBody@bb93a25
> D: 2020-10-19 15:01:03:294(LogBuilder.kt:33) .logHttp()[Network, response] [header] (Date, Mon, 19 Oct 2020 13:01:03 GMT)
> D: 2020-10-19 15:01:03:295(LogBuilder.kt:33) .logHttp()[Network, response] [info] Code: 207  Message: Multi-Status IsSuccessful: true
> D: 2020-10-19 15:01:03:295(LogBuilder.kt:33) .logHttp()[Network, response] [header] (Server, Apache)
> D: 2020-10-19 15:01:03:296(LogBuilder.kt:33) .logHttp()[Network, response] [header] (X-Content-Type-Options, nosniff)
> D: 2020-10-19 15:01:03:296(LogBuilder.kt:33) .logHttp()[Network, response] [header] (Date, Mon, 19 Oct 2020 13:01:03 GMT)
> D: 2020-10-19 15:01:03:297(LogBuilder.kt:33) .logHttp()[Network, response] [header] (Server, Apache)
> D: 2020-10-19 15:01:03:297(LogBuilder.kt:33) .logHttp()[Network, response] [header] (X-XSS-Protection, 1; mode=block)
> D: 2020-10-19 15:01:03:299(LogBuilder.kt:33) .logHttp()[Network, response] [header] (X-Content-Type-Options, nosniff)
> D: 2020-10-19 15:01:03:299(LogBuilder.kt:33) .logHttp()[Network, response] [header] (X-Robots-Tag, none)
> D: 2020-10-19 15:01:03:300(LogBuilder.kt:33) .logHttp()[Network, response] [header] (X-XSS-Protection, 1; mode=block)
> D: 2020-10-19 15:01:03:300(LogBuilder.kt:33) .logHttp()[Network, response] [header] (X-Frame-Options, SAMEORIGIN)
> D: 2020-10-19 15:01:03:301(LogBuilder.kt:33) .logHttp()[Network, response] [header] (X-Robots-Tag, none)
> D: 2020-10-19 15:01:03:301(LogBuilder.kt:33) .logHttp()[Network, response] [header] (X-Download-Options, noopen)
> D: 2020-10-19 15:01:03:302(LogBuilder.kt:33) .logHttp()[Network, response] [header] (X-Permitted-Cross-Domain-Policies, none)
> D: 2020-10-19 15:01:03:302(LogBuilder.kt:33) .logHttp()[Network, response] [header] (X-Frame-Options, SAMEORIGIN)
> D: 2020-10-19 15:01:03:303(LogBuilder.kt:33) .logHttp()[Network, response] [header] (X-Download-Options, noopen)
> D: 2020-10-19 15:01:03:303(LogBuilder.kt:33) .logHttp()[Network, response] [header] (Expires, Thu, 19 Nov 1981 08:52:00 GMT)
> D: 2020-10-19 15:01:03:304(LogBuilder.kt:33) .logHttp()[Network, response] [header] (Cache-Control, no-cache, no-store, must-revalidate)
> D: 2020-10-19 15:01:03:304(LogBuilder.kt:33) .logHttp()[Network, response] [header] (X-Permitted-Cross-Domain-Policies, none)
> D: 2020-10-19 15:01:03:306(LogBuilder.kt:33) .logHttp()[Network, response] [header] (Expires, Thu, 19 Nov 1981 08:52:00 GMT)
> D: 2020-10-19 15:01:03:308(LogBuilder.kt:33) .logHttp()[Network, response] [header] (Pragma, no-cache)
> D: 2020-10-19 15:01:03:308(LogBuilder.kt:33) .logHttp()[Network, response] [header] (Cache-Control, no-store, no-cache, must-revalidate)
> D: 2020-10-19 15:01:03:309(LogBuilder.kt:33) .logHttp()[Network, response] [header] (Pragma, no-cache)
> D: 2020-10-19 15:01:03:309(LogBuilder.kt:33) .logHttp()[Network, response] [header] (Content-Security-Policy, default-src 'none';manifest-src 'self';script-src 'self' 'unsafe-eval';style-src 'self' 'unsafe-inline';img-src 'self' data: blob:;font-src 'self';connect-src 'self';media-src 'self')
> D: 2020-10-19 15:01:03:311(LogBuilder.kt:33) .logHttp()[Network, response] [header] (Content-Security-Policy, default-src 'none';)
> D: 2020-10-19 15:01:03:311(LogBuilder.kt:33) .logHttp()[Network, response] [header] (Content-Length, 156)
> D: 2020-10-19 15:01:03:312(LogBuilder.kt:33) .logHttp()[Network, response] [header] (Keep-Alive, timeout=5, max=100)
> D: 2020-10-19 15:01:03:312(LogBuilder.kt:33) .logHttp()[Network, response] [header] (Vary, Brief,Prefer)
> D: 2020-10-19 15:01:03:313(LogBuilder.kt:33) .logHttp()[Network, response] [header] (Connection, Keep-Alive)
> D: 2020-10-19 15:01:03:313(LogBuilder.kt:33) .logHttp()[Network, response] [header] (DAV, 1, 3, extended-mkcol, 2)
> D: 2020-10-19 15:01:03:314(LogBuilder.kt:33) .logHttp()[Network, response] [header] (Keep-Alive, timeout=5, max=99)
> D: 2020-10-19 15:01:03:314(LogBuilder.kt:33) .logHttp()[Network, response] [header] (Content-Type, application/json; charset=utf-8)
> D: 2020-10-19 15:01:03:315(LogBuilder.kt:33) .logHttp()[Network, response] [header] (Connection, Keep-Alive)
> D: 2020-10-19 15:01:03:316(LogBuilder.kt:33) .logHttp()[Network, response] [body] okhttp3.internal.http.RealResponseBody@d3576a2
> D: 2020-10-19 15:01:03:317(LogBuilder.kt:33) .logHttp()[Network, response] [header] (Transfer-Encoding, chunked)
> ```
> 
> Here, all request headers and response headers are mixed. Probably this is a threading issue.... just to know if something can be done.

As you say, it is a threading issue. This happens when we execute several network operations at the same time. We could add an additional parameter to `[Network, request] [header]`, for example the request id: `[Network, request] [header][ccafec0a-bd8b-442d-b806-912385598035]` What do you think?

abelgardep commented:
> ### (2)
> Logging requests and reponses are duplicated, logging only the responses. For example:
> 
> 
> the set of headers marked as `[Network, request]` is the same as `[Network, response]`

Got it... 🤔

abelgardep commented:
> ### (1)
> Body requests and responses are not logged. This is the text logged instead:
> 
> ```
> okhttp3.internal.http.RealResponseBody@790056e
> ```

Two potential problems here:
1. Large files may produce OutOfMemoryError
2. Body can only be consumed once. In new architecture operations we try to log the body. For example in the GetRemoteUserInfoOperation. If something goes wrong with the body, we will see in the operation itself. The interceptor will help us to check headers, status, info, urls etc

`D/(GetRemoteUserInfoOperation.kt:58) .run(): Successful response {"ocs":{"meta":{"status":"ok","statuscode":200,"message":"OK","totalitems":"","itemsperpage":""},"data":{"id":"demo","display-name":"demo","email":null}}}`

abelgardep commented:
> ### (4)
> 1. Build a release apk
> 2. Add an account and enter in `Settings` -> Log section is hidden
> 3. Tap 5 times on build number to unlock
> 
> Current: Log section does not appear. You have to go out and enter again in `Settings` to make it visible
> Expected: Live appearance (is it feasible? in the `master` version, `Logs`are displayed just after the 5th click )

Fixed (2) and (4) @jesmrec

michaelstingl commented:
> * Large files may produce OutOfMemoryError

- Desktop and iOS doesn't log binary
- Desktop and iOS limit body to 1MB

(from https://github.com/owncloud/android/issues/2887)

jesmrec commented:
(1) -> in case we limit the binary logs in the body as @michaelstingl commented above, we have the second problem about the unique consumption of the body. Is there some room for improvement @abelgardep ?

(2) -> Now, requests and responses are correctly separated in logs. But, in case of the requests, i notice some missing headers. I guess this is because the lack of AppAuth support (to do). Headers are: `Cookies`, `Host`, `Connection`. Am i right?

(3) -> About including the Request-ID in every line... i'm not sure. Probably this is too much info and makes the lines quite long. But, no idea about any other solution. CC @michaelstingl 

(4) -> fixed

abelgardep commented:
> (1) -> in case we limit the binary logs in the body as @michaelstingl commented above, we have the second problem about the unique consumption of the body. Is there some room for improvement @abelgardep ?

First try to log not binary bodies. Ready to test.

> (2) -> Now, requests and responses are correctly separated in logs. But, in case of the requests, i notice some missing headers. I guess this is because the lack of AppAuth support (to do). Headers are: `Cookies`, `Host`, `Connection`. Am i right?

My bad, i added it as interceptor but not NetworkInterceptor. And AFAIK, cookies are added after interceptor, but before networkInterceptor. So it should work right now. Ready to test.

> (3) -> About including the Request-ID in every line... i'm not sure. Probably this is too much info and makes the lines quite long. But, no idea about any other solution. CC @michaelstingl

Added a new field: `"[Network, response] [header] [requestId] description"` We could reorder fields if needed: Maybe `"[Network, requestId] [header] [response] description"` fits better. Ready to test.

jesmrec commented:
### (5)

A small detail:

Body is now logged but, in the response body, the log is also marked as:

`[Network, request] [body]`

jesmrec commented:
### (6)

1. Enable option of logging network requests and responses
2. Upload a file (medium size ~50MB)

Current: Upload finishes with a `Unknown error` and file is not uploaded
Expected: Upload finishes

Could be related: https://github.com/owncloud/android/issues/2752

abelgardep commented:
I have been researching a little bit about the (6). We received `ProtocolException: expected 126311 bytes but received 126976` at the end of the upload. Problem fixed without logging chunks request body.

abelgardep commented:
At the same time, i was checking the logs impact in the chunks upload. Currently we have verbose logs in chunking:
```
com.owncloud.android.debug V/(ChunkFromFileRequestBody.java:91) .writeTo(): Read 4096 bytes from file channel to java.nio.HeapByteBuffer[pos=4096 lim=4096 cap=4096]
com.owncloud.android.debug V/(ChunkFromFileRequestBody.java:97) .writeTo(): Write 4096 bytes to sink buffer with size 0
com.owncloud.android.debug V/(ChunkFromFileRequestBody.java:87) .writeTo(): Sink buffer size: 0
```
Those lines are repeated when we want to upload a file with chunks each time we read 4096 bytes. For large files those are a lot of logs and they affect performance. I have measured time uploading a 12.6Mb file with logs and without them. 
- With logs it takes 53653 milliseconds to upload the file.
- Without logs it takes 4927 milliseconds to upload the file.

From my point those logs are noise and we could get rid of them. What do you think @michaelstingl @jesmrec @theScrabi

jesmrec commented:
From my side, it's OK to get rid of the binary file logging. They don't give me sensitive information

michaelstingl commented:
> From my point those logs are noise and we could get rid of them. What do you think @michaelstingl @jesmrec @theScrabi

Yeah, reading data blocks from file system is not that interesting for the logs. Maybe only the beginning and the end. But I'm still interested in the HTTP requests and responses from the full chunks.

abelgardep commented:
> > From my point those logs are noise and we could get rid of them. What do you think @michaelstingl @jesmrec @theScrabi
> 
> Yeah, reading data blocks from file system is not that interesting for the logs. Maybe only the beginning and the end. But I'm still interested in the HTTP requests and responses from the full chunks.

Of course, we will log every HTTP request and response, but i think that those read/write buffer logs have too much impact in uploads time. And from my point of view, they are not helpful at all.

jesmrec commented:
### (7)

A couple of considerations about body responses

PROPFIND responses:

```
D: 2020-10-27 14:14:28:598(LogBuilder.kt:34) .logHttp()[Network, response] [body] [2ae126de-6671-46e5-b33a-1f6f585242cc] --> Body start for request
D: 2020-10-27 14:14:28:600(LogBuilder.kt:34) .logHttp()[Network, response] [body] [2ae126de-6671-46e5-b33a-1f6f585242cc] <?xml version="1.0"?>
D: 2020-10-27 14:14:28:602(LogBuilder.kt:34) .logHttp()[Network, response] [body] [2ae126de-6671-46e5-b33a-1f6f585242cc] <-- Body end for request -- Omitted: 0 bytes
```

logged response only contains the first line `<?xml version="1.0"?>` . Is this intended?

OTOH,  an small typo:  `Body start for request` -> `Body start for response`

abelgardep commented:
(5) Ready to test
(7) Typo fixed, but i wasn't able to reproduce the problem.

jesmrec commented:
Everything is fixed, so from my side it is OK to go with the first iteration. Next: https://github.com/owncloud/android/issues/3009

rashato commented:
Is the issue raised in https://github.com/owncloud/android/issues/2752 corrected? Is the latest version available in google play? I'm not used to git and I'm afraid not to be able to commit.

jesmrec commented:
This is not totally related to #2752 but it can help to fix it. For the moment, the fix to #2752 is neither done nor in Google Play. Stay tunned.
